send/m365_sharepoint

```diff
@@ -0,0 +1,611 @@
+	post_data = {
+		'client_id': client_id,
+		'client_secret': client_secret,
+		'scope': 'https://graph.microsoft.com/.default',
+		'grant_type': 'client_credentials'
+	}
+	response = requests.post(auth_endpoint, post_data)
+	print_debug(f'Response from auth endpoint ended with status code {response.status_code}')
+	if response.ok:
+		response_data = response.json()
+		global access_token
+		access_token = response_data['access_token']
+	else:
+		send_lib.die_with_error("Communication with authorization endpoint failed.")
+
+
+def generate_random_password(length=16):
+	"""
+	Generates random password that should comply with password policies specified here:
+	https://learn.microsoft.com/en-us/azure/active-directory/authentication/concept-sspr-policy#azure-ad-password-policies
+	"""
+	lowercase_letters = string.ascii_lowercase
+	uppercase_letters = string.ascii_uppercase
+	digits = string.digits
+	special_characters = '!@#$%^&*()_-+=<>?'
+
+	password = [
+		secrets.choice(lowercase_letters),
+		secrets.choice(uppercase_letters),
+		secrets.choice(digits),
+		secrets.choice(special_characters)
+	]
+
+	remaining_length = length - len(password)
+	all_characters = lowercase_letters + uppercase_letters + digits + special_characters
+	password.extend(secrets.choice(all_characters) for _ in range(remaining_length))
+	secrets.SystemRandom().shuffle(password)
+	generated_password = ''.join(password)
+
+	return generated_password
+
+
+def get_our_internal_users(our_data):
+	""" Returns all members of resources marked as m365 license groups """
+	link_ids = set()
+	for group in our_data["groups"]:
+		if group[islicencegroup_attr_name]:
+			for member in group["members"]:
+				link_ids.add(member["link_id"])
+	internals = list(filter(lambda u: u["link_id"] in link_ids, our_data["users"]))
+	return internals
+
+
+def get_immutable_id(our_user):
+	""" Creates immutable ID - it has to be unique, including restorable objects """
+	# return base64.b64encode(f"{our_user[login_attr_name]}@{scope}".encode()).decode('ascii')
+	return f'{our_user[login_attr_name]}@{SCOPE}'
+
+
+def call_api(url: str, params=None, type='GET', retry_no=0):
+	"""
+	Calls the API. Handles slowing requests down if limits are hit and renews access token if needed.
+	@return: response data
+	"""
+	if retry_no > 3:
+		send_lib.die_with_error("Exceeded max number of retries.")
+	if type == 'GET':
+		response = requests.get(url=url, params=params, headers={"Authorization": "Bearer " + access_token},
+								timeout=TIMEOUT)
+	elif type == 'DELETE':
+		response = requests.delete(url=url, json=params, headers={"Authorization": "Bearer " + access_token},
+								   timeout=TIMEOUT)
+	elif type == 'PATCH':
+		response = requests.patch(url=url, json=params, headers={"Authorization": "Bearer " + access_token},
+								  timeout=TIMEOUT)
+	elif type == 'POST':
+		response = requests.post(url=url, json=params, headers={"Authorization": "Bearer " + access_token},
+								 timeout=TIMEOUT)
```

Comment:
Suggestion only: The header can be created once and used for all the HTTP operations.